src/encode.rs

```diff
@@ -0,0 +1,296 @@
+    ///
+    /// Encodes the key with the header:
+    /// `-----BEGIN PUBLIC KEY-----`
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use rsa::{RSAPrivateKey, RSAPublicKey};
+    /// use rand::rngs::OsRng;
+    ///
+    /// let mut rng = OsRng;
+    /// let bits = 2048;
+    /// let private_key = RSAPrivateKey::new(&mut rng, bits).expect("failed to generate a key");
+    /// let public_key = RSAPublicKey::from(&private_key);
+    ///
+    /// let _ = public_key.to_pem_pkcs8();
+    /// ```
+    pub fn to_pem_pkcs8(&self) -> Result<String> {
+        let pem = pem::Pem {
+            tag: String::from("PUBLIC KEY"),
+            contents: self.to_pkcs8()?,
+        };
+        Ok(pem::encode(&pem))
+    }
+}
+
+fn to_bigint(value: &crate::BigUint) -> simple_asn1::BigInt {
+    // TODO can be switched if simple_asn1 BigInt type is updated
+    // This is not very clean because of the exports available from simple_asn1
+    simple_asn1::BigInt::from_signed_bytes_le(&value.to_bigint().unwrap().to_signed_bytes_le())
+}
+
+impl RSAPrivateKey {
+    /// Encodes an RSA Private key to into `PKCS1` bytes.
+    ///
+    /// This data will be `base64` encoded which would be used
+    /// following a `-----BEGIN RSA PRIVATE KEY-----` header.
+    ///
+    /// <https://tls.mbed.org/kb/cryptography/asn1-key-structures-in-der-and-pem>
+    pub fn to_pkcs1(&self) -> Result<Vec<u8>> {
+        // TODO should version be changed to anything?
+        let version = ASN1Block::Integer(0, to_bigint(&BigUint::zero()));
+        let n = ASN1Block::Integer(0, to_bigint(&self.n()));
+        let e = ASN1Block::Integer(0, to_bigint(&self.e()));
+        let d = ASN1Block::Integer(0, to_bigint(&self.d));
+        let mut blocks = vec![version, n, e, d];
+
+        // Encode primes
+        blocks.extend(
+            self.primes
+                .iter()
+                .map(|p| ASN1Block::Integer(0, to_bigint(p))),
+        );
+        to_der(&ASN1Block::Sequence(0, blocks)).map_err(|e| Error::EncodeError {
+            reason: format!("failed to encode ASN.1 sequence of blocks: {}", e),
+        })
+    }
+
+    /// Encodes an RSA Private key to into `PKCS8` bytes.
+    ///
+    /// This data will be `base64` encoded which would be used
+    /// following a `-----BEGIN PRIVATE KEY-----` header.
+    ///
+    /// <https://tls.mbed.org/kb/cryptography/asn1-key-structures-in-der-and-pem>
+    pub fn to_pkcs8(&self) -> Result<Vec<u8>> {
+        let version = ASN1Block::Integer(0, to_bigint(&BigUint::zero()));
+        let oid = ASN1Block::ObjectIdentifier(0, rsa_oid());
+        let alg = ASN1Block::Sequence(0, vec![oid]);
+        let octet_string = ASN1Block::OctetString(0, self.to_pkcs1()?);
+        let blocks = vec![version, alg, octet_string];
+
+        to_der(&ASN1Block::Sequence(0, blocks)).map_err(|e| Error::EncodeError {
+            reason: format!("failed to encode ASN.1 sequence of blocks: {}", e),
+        })
+    }
+}
+
+impl RSAPublicKey {
```

Comment:
This entire `impl` block could instead be an extension trait (e.g. `trait PublicKeyEncoding : PublicKey`) that is auto-implemented for all types that implement `PublicKey`.